File: bindings/python/src/espn_polymarket_map/parse.py
"""Parsing raw ESPN scoreboard events and Polymarket gamma events into typed models."""

from __future__ import annotations

import json
from typing import Any, List, Optional

from .models import EspnGame, EspnSide, PmEvent, PmMarket

__all__ = ["parse_espn_event", "parse_pm_event"]


def _s(v: Any, key: str) -> str:
    """Read ``v[key]`` as a string, returning ``""`` if missing or not a string."""
    if isinstance(v, dict):
        val = v.get(key)
        if isinstance(val, str):
            return val
    return ""


def _value_to_string(v: Any) -> str:
    """Mirror serde_json's ``Value::to_string`` for non-string elements.

    Strings pass through verbatim; any other JSON value is serialized compactly.
    """
    if isinstance(v, str):
        return v
    if v is True:
        return "true"
    if v is False:
        return "false"
    if v is None:
        return "null"
    if isinstance(v, (int, float)):
        # serde_json renders integers without a decimal point and floats minimally;
        # json.dumps matches for the integer/float string forms encountered here.
        return json.dumps(v)
    return json.dumps(v, separators=(",", ":"))


def _string_array(v: Any) -> List[str]:
    """Parse a Polymarket ``outcomes`` / ``outcomePrices`` / ``clobTokenIds`` field.

    Gamma encodes these as a JSON-string-of-array (e.g. ``"[\\"Yes\\", \\"No\\"]"``).
    Also tolerates a real array.
    """
    if isinstance(v, str):
        try:
            arr = json.loads(v)
        except (ValueError, TypeError):
            return []
        if isinstance(arr, list):
            return [_value_to_string(x) for x in arr]
        return []
    if isinstance(v, list):
        return [_value_to_string(x) for x in v]
    return []


def parse_espn_event(json_obj: Any, league: str) -> Optional[EspnGame]:
    """Parse one ESPN scoreboard ``event`` object into an :class:`EspnGame`."""
    comp = json_obj
    if isinstance(json_obj, dict):
        comps = json_obj.get("competitions")
        if isinstance(comps, list) and comps:
            comp = comps[0]

    kickoff = _s(comp, "date")
    if not kickoff:
        kickoff = _s(json_obj, "date")

    competitors = comp.get("competitors") if isinstance(comp, dict) else None
    if not isinstance(competitors, list) or len(competitors) < 2:
        return None

    def side(c: Any) -> EspnSide:
        # A side is a team (most sports) or an athlete (tennis, MMA).
        t = c.get("team") if isinstance(c, dict) else None
        if not isinstance(t, dict):
            t = c.get("athlete") if isinstance(c, dict) else None
        if not isinstance(t, dict):
            t = c
        return EspnSide(
            abbr=_s(t, "abbreviation"),
            display_name=_s(t, "displayName"),
            short_name=_s(t, "shortDisplayName") or _s(t, "shortName"),
            location=_s(t, "location"),
            nickname=_s(t, "name") or _s(t, "lastName"),
        )

    def find(home_away: str) -> Any:
        for c in competitors:
            if _s(c, "homeAway") == home_away:
                return c
        return None

    h, a = find("home"), find("away")
    if h is None or a is None:
        # Athlete head-to-head events carry no home/away — fall back to competitor order.
        a, h = competitors[0], competitors[1]
    home = side(h)
    away = side(a)

    return EspnGame(
        league=league,
        espn_event_id=_s(json_obj, "id"),
        kickoff_utc=kickoff,
        home=home,
        away=away,
    )


def parse_pm_event(json_obj: Any) -> PmEvent:
    """Parse a Polymarket gamma ``event`` object into a :class:`PmEvent`."""
    markets: List[PmMarket] = []
    raw_markets = json_obj.get("markets") if isinstance(json_obj, dict) else None
    if isinstance(raw_markets, list):
        for m in raw_markets:
            smt_raw = m.get("sportsMarketType") if isinstance(m, dict) else None
            smt = smt_raw if isinstance(smt_raw, str) else None
            markets.append(
                PmMarket(
                    slug=_s(m, "slug"),
                    question=_s(m, "question"),
                    group_item_title=_s(m, "groupItemTitle"),
                    sports_market_type=smt,
                    outcomes=_string_array(m.get("outcomes") if isinstance(m, dict) else None),
                    outcome_prices=_string_array(
                        m.get("outcomePrices") if isinstance(m, dict) else None
                    ),
                    clob_token_ids=_string_array(
                        m.get("clobTokenIds") if isinstance(m, dict) else None
                    ),
                )
            )

    return PmEvent(
        slug=_s(json_obj, "slug"),
        title=_s(json_obj, "title"),
        markets=markets,
    )

File: core-rs/src/parse.rs
//! Parsing raw ESPN scoreboard events and Polymarket gamma events into typed models.

use crate::models::{EspnGame, EspnSide, PmEvent, PmMarket};
use serde_json::Value;

fn s(v: &Value, key: &str) -> String {
    v.get(key).and_then(Value::as_str).unwrap_or("").to_string()
}

/// Parse a Polymarket `outcomes`/`outcomePrices`/`clobTokenIds` field, which gamma encodes
/// as a JSON-string-of-array (e.g. `"[\"Yes\", \"No\"]"`). Also tolerates a real array.
fn string_array(v: Option<&Value>) -> Vec<String> {
    match v {
        Some(Value::String(s)) => serde_json::from_str::<Vec<Value>>(s)
            .map(|arr| arr.iter().map(value_to_string).collect())
            .unwrap_or_default(),
        Some(Value::Array(arr)) => arr.iter().map(value_to_string).collect(),
        _ => Vec::new(),
    }
}

fn value_to_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

/// Parse one ESPN scoreboard `event` object into an [`EspnGame`] for the given league.
pub fn parse_espn_event(json: &Value, league: &str) -> Option<EspnGame> {
    let comp = json
        .get("competitions")
        .and_then(|c| c.get(0))
        .unwrap_or(json);
    let kickoff = {
        let c = s(comp, "date");
        if c.is_empty() {
            s(json, "date")
        } else {
            c
        }
    };
    let competitors = comp.get("competitors").and_then(Value::as_array)?;
    if competitors.len() < 2 {
        return None;
    }

    // A side is a team (most sports) or an athlete (tennis, MMA); read whichever is present.
    let side = |c: &Value| -> EspnSide {
        let t = c.get("team").or_else(|| c.get("athlete")).unwrap_or(c);
        EspnSide {
            abbr: s(t, "abbreviation"),
            display_name: s(t, "displayName"),
            short_name: {
                let sd = s(t, "shortDisplayName");
                if sd.is_empty() {
                    s(t, "shortName")
                } else {
                    sd
                }
            },
            location: s(t, "location"),
            nickname: {
                let n = s(t, "name");
                if n.is_empty() {
                    s(t, "lastName")
                } else {
                    n
                }
            },
        }
    };

    // Use explicit home/away when present; otherwise fall back to competitor order
    // (athlete head-to-head events have no home/away).
    let find = |ha: &str| competitors.iter().find(|c| s(c, "homeAway") == ha);
    let (home, away) = match (find("home"), find("away")) {
        (Some(h), Some(a)) => (h, a),
        _ => (&competitors[1], &competitors[0]),
    };

    Some(EspnGame {
        league: league.to_string(),
        espn_event_id: s(json, "id"),
        kickoff_utc: kickoff,
        home: side(home),
        away: side(away),
    })
}

/// Parse a Polymarket gamma `event` object into a [`PmEvent`].
pub fn parse_pm_event(json: &Value) -> PmEvent {
    let markets = json
        .get("markets")
        .and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .map(|m| PmMarket {
                    slug: s(m, "slug"),
                    question: s(m, "question"),
                    group_item_title: s(m, "groupItemTitle"),
                    sports_market_type: m
                        .get("sportsMarketType")
                        .and_then(Value::as_str)
                        .map(str::to_string),
                    outcomes: string_array(m.get("outcomes")),
                    outcome_prices: string_array(m.get("outcomePrices")),
                    clob_token_ids: string_array(m.get("clobTokenIds")),
                })
                .collect()
        })
        .unwrap_or_default();

    PmEvent {
        slug: s(json, "slug"),
        title: s(json, "title"),
        markets,
    }
}
